server/message_dispatch.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"strings"
+	"sync"
+)
+
+type MessageDispatch struct {
+	input       chan string
+	output      map[string][]chan string
```

Comment:
We actually had a discussion about this and couldn't figure out a good way to do it using conds. It's hard to prevent notification-misses and subscription-misses using Conds without a memory leak - basically, we can't know when to "cleanup" our Conds without multiple mutexes. That's why we implemented our own broadcasting mechanism. Channels not being buffered is definitely a problem though.